client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java

```diff
@@ -102,64 +100,32 @@ public long getOffset() {
      */
     @Override
     @Synchronized
-    public ByteBuffer read(long timeout) throws EndOfSegmentException, SegmentTruncatedException {
+    public int read(ByteBuffer toFill, long timeout) throws EndOfSegmentException, SegmentTruncatedException {
         Exceptions.checkNotClosed(asyncInput.isClosed(), this);
-        long originalOffset = offset;
-        long traceId = LoggerHelpers.traceEnter(log, "read", getSegmentId(), originalOffset, timeout);
-        boolean success = false;
-        try {
-            ByteBuffer result = readEventData(timeout);
-            success = true;
-            return result;
-        } finally {
-            LoggerHelpers.traceLeave(log, "read", traceId, getSegmentId(), originalOffset, timeout);
-            if (!success) {
-                outstandingRequest = null;
-                offset = originalOffset;
-                buffer.clear();
-            }
-        }
-    }
-
-    private ByteBuffer readEventData(long timeout) throws EndOfSegmentException, SegmentTruncatedException {
         if (this.offset >= this.endOffset) {
             log.debug("All events up to the configured end offset:{} have been read", endOffset);
             throw new EndOfSegmentException(END_OFFSET_REACHED);
         }
-        fillBuffer();
+        if (outstandingRequest == null) {
+            fillBuffer();
+        }
         if (receivedTruncated) {
             throw new SegmentTruncatedException();
         }
-        while (buffer.dataAvailable() < WireCommands.TYPE_PLUS_LENGTH_SIZE) {
+        while (buffer.dataAvailable() == 0) {
```

Comment:
I don't understand this change, it seems to be more conservative, and the reason isn't clear.

Reply:
This class is no longer dealing in events. It is dealing with just bytes and the wrapping API is now dealing in events. As such it is "ok" for this API to read less than 8 bytes at a time.